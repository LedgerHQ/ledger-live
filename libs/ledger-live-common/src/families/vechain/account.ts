import type { Account } from "@ledgerhq/types-live";
import { getAccountUnit } from "../../account";
import { formatCurrencyUnit } from "../../currencies";

function formatAccountSpecifics(account: Account): string {
  const unit = getAccountUnit(account);
  const formatConfig = {
    disableRounding: true,
    alwaysShowSign: false,
    showCode: true,
  };

  let str = " ";

  str += formatCurrencyUnit(unit, account.spendableBalance, formatConfig) + " spendable. ";

  return str;
}

export default {
  formatAccountSpecifics,
};
