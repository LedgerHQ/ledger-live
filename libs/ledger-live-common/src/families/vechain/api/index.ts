export { getAccount, getOperations } from "./sdk";
export * from "./sdk";
