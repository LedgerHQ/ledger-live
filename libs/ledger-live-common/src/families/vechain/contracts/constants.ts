export const TransferEventSignature =
  "0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef";

export const VTHO_ADDRESS = "0x0000000000000000000000000000456e65726779";

// params.sol - builtin contract (https://www.vechain.org/whitepaper/#bit_v48i3)
export const PARAMS_ADDRESS = "0x0000000000000000000000000000506172616d73";
export const REWARD_RATIO_KEY =
  "0x00000000000000000000000000000000000000007265776172642d726174696f";
export const BASE_GAS_PRICE_KEY =
  "0x000000000000000000000000000000000000626173652d6761732d7072696365";
