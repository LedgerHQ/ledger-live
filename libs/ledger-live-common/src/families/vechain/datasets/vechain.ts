import { AccountRaw } from "@ledgerhq/types-live";

// NOTE: account information is fetched from the blockchain so this is just integration for those information
// so you you should not rely on that, here the important part are ids and index

export const vechain1: AccountRaw = {
  id: "js:2:vechain:0x0fe6688548f0C303932bB197B0A96034f1d74dba:vechain",
  seedIdentifier: "0x0fe6688548f0C303932bB197B0A96034f1d74dba",
  name: "Vechain 1",
  derivationMode: "vechain",
  index: 0,
  freshAddress: "0x0fe6688548f0C303932bB197B0A96034f1d74dba",
  freshAddressPath: "44'/818'/0'/0/0",
  blockHeight: 15156315,
  operations: [],
  pendingOperations: [],
  currencyId: "vechain",
  feesCurrencyId: "vechain/vip180/vtho",
  lastSyncDate: "2023-04-21T14:22:48.395Z",
  balance: "10000000000000000000",
  subAccounts: [
    {
      type: "TokenAccountRaw",
      id: "js:2:vechain:0x0fe6688548f0C303932bB197B0A96034f1d74dba:vechain+vechain%2Fvip180%2Fvtho",
      parentId: "js:2:vechain:0x0fe6688548f0C303932bB197B0A96034f1d74dba:vechain",
      tokenId: "vechain/vip180/vtho",
      balance: "10000000000000000000",
      operations: [],
      pendingOperations: [],
    },
  ],
  swapHistory: [],
};

export const vechain2: AccountRaw = {
  id: "js:2:vechain:0x02961B92B8D20A4ea12f1f1CeFA74Dd7B4355A86:vechain",
  seedIdentifier: "0x02961B92B8D20A4ea12f1f1CeFA74Dd7B4355A86",
  name: "Vechain 2",
  derivationMode: "vechain",
  index: 1,
  freshAddress: "0x02961B92B8D20A4ea12f1f1CeFA74Dd7B4355A86",
  freshAddressPath: "44'/818'/0'/0/1",
  blockHeight: 15156315,
  operations: [],
  pendingOperations: [],
  currencyId: "vechain",
  feesCurrencyId: "vechain/vip180/vtho",
  lastSyncDate: "2023-04-21T14:22:48.395Z",
  balance: "10000000000000000000",
  subAccounts: [
    {
      type: "TokenAccountRaw",
      id: "js:2:vechain:0x02961B92B8D20A4ea12f1f1CeFA74Dd7B4355A86:vechain+vechain%2Fvip180%2Fvtho",
      parentId: "js:2:vechain:0x02961B92B8D20A4ea12f1f1CeFA74Dd7B4355A86:vechain",
      tokenId: "vechain/vip180/vtho",
      balance: "0",
      operations: [],
      pendingOperations: [],
    },
  ],
  swapHistory: [],
};

// account with vtho but no vet
export const vechain3: AccountRaw = {
  id: "js:2:vechain:0xC0dFC490f8fba6A573C55a68dd9023f999ccaDA0:vechain",
  seedIdentifier: "0xC0dFC490f8fba6A573C55a68dd9023f999ccaDA0",
  name: "Vechain 3",
  derivationMode: "vechain",
  index: 1,
  freshAddress: "0xC0dFC490f8fba6A573C55a68dd9023f999ccaDA0",
  freshAddressPath: "44'/818'/0'/0/0",
  blockHeight: 15156315,
  operations: [],
  pendingOperations: [],
  currencyId: "vechain",
  feesCurrencyId: "vechain/vip180/vtho",
  lastSyncDate: "2023-11-13T13:55:32.004Z",
  balance: "0",
  subAccounts: [
    {
      type: "TokenAccountRaw",
      id: "js:2:vechain:0xC0dFC490f8fba6A573C55a68dd9023f999ccaDA0:vechain+vechain%2Fvip180%2Fvtho",
      parentId: "js:2:vechain:0xC0dFC490f8fba6A573C55a68dd9023f999ccaDA0:vechain",
      tokenId: "vechain/vip180/vtho",
      balance: "1000000000000000",
      operations: [],
      pendingOperations: [],
    },
  ],
  swapHistory: [],
};
