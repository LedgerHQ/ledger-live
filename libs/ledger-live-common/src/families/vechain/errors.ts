import { createCustomErrorClass } from "@ledgerhq/errors";

export const NotEnoughVTHO = createCustomErrorClass("NotEnoughVTHO");
export const MustBeVechain = createCustomErrorClass("MustBeVechain");
export const ImpossibleToCalculateAmountAndFees = createCustomErrorClass(
  "ImpossibleToCalculateAmountAndFees",
);
