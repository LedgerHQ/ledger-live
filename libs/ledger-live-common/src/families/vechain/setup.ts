// Goal of this file is to inject all necessary device/signer dependency to coin-modules

import { createBridges } from "@ledgerhq/coin-vechain/index";
import makeCliTools from "@ledgerhq/coin-vechain/test/cli";
import vechainResolver, { signMessage } from "@ledgerhq/coin-vechain/signer/index";
import { Transaction, VechainSigner } from "./types";
import Transport from "@ledgerhq/hw-transport";
import type { Bridge } from "@ledgerhq/types-live";
import Vet from "@ledgerhq/hw-app-vet";
import { CreateSigner, createResolver, executeWithSigner } from "../../bridge/setup";
import type { Resolver } from "../../hw/getAddress/types";

const createSigner: CreateSigner<VechainSigner> = (transport: Transport) => new Vet(transport);

const bridge: Bridge<Transaction> = createBridges(executeWithSigner(createSigner));

const messageSigner = {
  signMessage,
};

const resolver: Resolver = createResolver(createSigner, vechainResolver);

const cliTools = makeCliTools();

export { bridge, cliTools, messageSigner, resolver };
