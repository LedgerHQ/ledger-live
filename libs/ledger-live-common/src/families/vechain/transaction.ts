import type { Transaction, TransactionRaw, TransactionStatus, TransactionStatusRaw } from "./types";
import {
  formatTransactionStatusCommon as formatTransactionStatus,
  fromTransactionCommonRaw,
  fromTransactionStatusRawCommon,
  toTransactionCommonRaw,
  toTransactionStatusRawCommon,
} from "@ledgerhq/coin-framework/transaction/common";
import type { Account } from "@ledgerhq/types-live";
import { getAccountUnit } from "../../account";
import { formatCurrencyUnit } from "../../currencies";

export const formatTransaction = (t: Transaction, account: Account): string => {
  const { amount, recipient, useAllAmount } = t;
  let displayedAmount: string;
  if (useAllAmount) {
    displayedAmount = "MAX";
  } else if (amount.isZero()) {
    displayedAmount = "";
  } else {
    displayedAmount =
      " " +
      formatCurrencyUnit(getAccountUnit(account), amount, {
        showCode: true,
        disableRounding: true,
      });
  }
  return `SEND ${displayedAmount} TO ${recipient}`;
};

export const fromTransactionRaw = (tr: TransactionRaw): Transaction => {
  const common = fromTransactionCommonRaw(tr);

  return {
    ...tr,
    ...common,
  };
};

export const toTransactionRaw = (t: Transaction): TransactionRaw => {
  const common = toTransactionCommonRaw(t);

  return {
    ...t,
    ...common,
  };
};

export const fromTransactionStatusRaw = (ts: TransactionStatusRaw): TransactionStatus => {
  const common = fromTransactionStatusRawCommon(ts);

  return {
    ...ts,
    ...common,
  };
};

export const toTransactionStatusRaw = (ts: TransactionStatus): TransactionStatusRaw => {
  const common = toTransactionStatusRawCommon(ts);

  return {
    ...ts,
    ...common,
  };
};

export default {
  formatTransaction,
  formatTransactionStatus,
  fromTransactionRaw,
  toTransactionRaw,
  fromTransactionStatusRaw,
  toTransactionStatusRaw,
};
