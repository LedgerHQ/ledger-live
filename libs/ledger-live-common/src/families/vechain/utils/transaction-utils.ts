import { DEFAULT_GAS_COEFFICIENT, HEX_PREFIX } from "../constants";
import crypto from "crypto";
import BigNumber from "bignumber.js";
import { Transaction as ThorTransaction } from "thor-devkit";
import params from "../contracts/abis/params";
import { BASE_GAS_PRICE_KEY, PARAMS_ADDRESS } from "../contracts/constants";
import { Query } from "../api/types";
import { query } from "../api/sdk";
import { Account, TokenAccount } from "@ledgerhq/types-live";
import { Transaction, TransactionInfo } from "../types";
import { isValid } from "./address-utils";
import { calculateClausesVet, calculateClausesVtho } from "../logic";
import { ImpossibleToCalculateAmountAndFees } from "../errors";

const GAS_COEFFICIENT = 15000;

/**
 * Generate a Unique ID to be used as a nonce
 * @returns a unique string
 */
export const generateNonce = (): string => {
  const randBuffer = crypto.randomBytes(Math.ceil(4));
  if (!randBuffer) throw Error("Failed to generate random hex");
  return `${HEX_PREFIX}${randBuffer.toString("hex").substring(0, 8)}`;
};

/**
 * Estimate the gas that will be used by the transaction.
 * @param transaction - The transaction to estimate the gas for
 * @returns an estimate of the gas usage
 */
export const estimateGas = async (t: Transaction): Promise<number> => {
  const intrinsicGas = ThorTransaction.intrinsicGas(t.body.clauses);

  const queryData: Query[] = [];

  t.body.clauses.forEach(c => {
    if (c.to) {
      queryData.push({
        to: c.to,
        data: c.data,
      });
    }
  });

  const response = await query(queryData);

  const execGas = response.reduce((sum, out) => sum + out.gasUsed, 0);

  return intrinsicGas + (execGas ? execGas + GAS_COEFFICIENT : 0);
};

const getBaseGasPrice = async (): Promise<string> => {
  const queryData: Query = {
    to: PARAMS_ADDRESS,
    data: params.get.encode(BASE_GAS_PRICE_KEY),
  };

  const response = await query([queryData]);

  // Expect 1 value
  if (response && response.length != 1) throw Error("Unexpected response received for query");

  return response[0].data;
};

/**
 * Calculate the fee in VTHO
 * @param gas - the gas used
 * @param gasPriceCoef - the gas price coefficient
 * @returns the fee in VTHO
 */
export const calculateFee = async (gas: BigNumber, gasPriceCoef: number): Promise<BigNumber> => {
  const baseGasPrice = await getBaseGasPrice();
  return new BigNumber(baseGasPrice).times(gasPriceCoef).idiv(255).plus(baseGasPrice).times(gas);
};

// Here there is a circular dependency between values, that is why we need the do-while loop
// dependencies are:
// useAllAmount: USER
// amount: useAllAmount & spendableBalance
// fees: amount
// spendableBalance: fees & balance
// balance: USER
// circular dependency is:
// amount -> spendableBalance -> fees -> amount

export const calculateTransactionInfo = async (
  account: Account,
  transaction: Transaction,
  fixedMaxTokenFees?: {
    estimatedGas: number;
    estimatedGasFees: BigNumber;
  },
): Promise<TransactionInfo> => {
  const { subAccounts } = account;
  const { amount: oldAmount, subAccountId, useAllAmount } = transaction;

  const tokenAccount =
    subAccountId && subAccounts
      ? (subAccounts.find(subAccount => {
          return subAccount.id === subAccountId;
        }) as TokenAccount)
      : undefined;
  const isTokenAccount = !!tokenAccount;

  let amount = oldAmount;
  let amountBackup;
  let tempTransaction = { ...transaction, amount };
  let balance = account.balance;
  let spendableBalance = account.balance;
  let maxEstimatedGasFees = new BigNumber(0);
  let maxEstimatedGas = 0;

  if (!amount.isNaN()) {
    const MAX_ITERATIONS = 5; // it should never reach more than 2 iterations, but just in case
    let iterations = 0;
    do {
      amountBackup = amount;

      const estimatedGasAndFees =
        fixedMaxTokenFees || (await calculateGasFees(tempTransaction, isTokenAccount));

      maxEstimatedGasFees = estimatedGasAndFees.estimatedGasFees;
      maxEstimatedGas = estimatedGasAndFees.estimatedGas;

      if (isTokenAccount && tokenAccount) {
        balance = tokenAccount.balance;
        spendableBalance = tokenAccount.balance.minus(maxEstimatedGasFees).gt(0)
          ? tokenAccount.balance.minus(maxEstimatedGasFees)
          : new BigNumber(0);
      } else {
        balance = account.balance;
        spendableBalance = account.balance;
      }

      amount = useAllAmount ? spendableBalance : oldAmount;

      tempTransaction = {
        ...tempTransaction,
        amount,
      };
      iterations++;
    } while (!amountBackup.isEqualTo(amount) && iterations < MAX_ITERATIONS);
    if (iterations === MAX_ITERATIONS) {
      throw new ImpossibleToCalculateAmountAndFees();
    }
  }

  return {
    isTokenAccount,
    amount,
    spendableBalance,
    balance,
    tokenAccount,
    estimatedFees: maxEstimatedGasFees.toString(),
    estimatedGas: maxEstimatedGas,
  };
};

export const calculateGasFees = async (
  transaction: Transaction,
  isTokenAccount: boolean,
): Promise<{
  estimatedGas: number;
  estimatedGasFees: BigNumber;
}> => {
  if (transaction.recipient && isValid(transaction.recipient)) {
    let clauses;
    if (isTokenAccount) {
      clauses = await calculateClausesVtho(transaction.recipient, transaction.amount);
    } else {
      clauses = await calculateClausesVet(transaction.recipient, transaction.amount);
    }
    const gas = await estimateGas({
      ...transaction,
      body: { ...transaction.body, clauses: clauses },
    });
    return {
      estimatedGas: gas,
      estimatedGasFees: await calculateFee(
        new BigNumber(gas),
        transaction.body.gasPriceCoef || DEFAULT_GAS_COEFFICIENT,
      ),
    };
  }
  return {
    estimatedGas: 0,
    estimatedGasFees: new BigNumber(0),
  };
};
