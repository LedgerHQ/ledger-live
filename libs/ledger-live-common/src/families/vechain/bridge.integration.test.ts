import "../../__tests__/test-helpers/setup";
import { testBridge } from "../../__tests__/test-helpers/bridge";
import type { AccountRaw, DatasetTest } from "@ledgerhq/types-live";
import type { Transaction } from "./types";
import { fromTransactionRaw } from "./transaction";
import { DEFAULT_GAS_COEFFICIENT, MAINNET_CHAIN_TAG } from "./constants";
import { vechain1, vechain3 } from "./datasets/vechain";
import { generateNonce } from "./utils/transaction-utils";
import BigNumber from "bignumber.js";
import {
  setSupportedCurrencies,
  listSupportedCurrencies,
  getCryptoCurrencyById,
} from "../../currencies";
import vechainScanAccounts1 from "./datasets/vechain.scanAccounts.1";
import { AmountRequired, NotEnoughBalance } from "@ledgerhq/errors";
import VIP180 from "./contracts/abis/VIP180";
import { CryptoCurrencyId } from "@ledgerhq/types-cryptoassets";
import { NotEnoughVTHO } from "./errors";

const listSupported = listSupportedCurrencies();
listSupported.push(getCryptoCurrencyById("vechain"));
setSupportedCurrencies(listSupported.map(c => c.id) as CryptoCurrencyId[]);

const dataset: DatasetTest<Transaction> = {
  implementations: ["js", "mock"],
  currencies: {
    vechain: {
      FIXME_ignoreAccountFields: ["balance", "spendableBalance", "estimateMaxSpendable"], // the balance depends on VTHO and it's earned without operations
      scanAccounts: vechainScanAccounts1,

      accounts: [
        {
          raw: vechain1 as AccountRaw,
          transactions: [
            {
              name: "Send VET",
              transaction: fromTransactionRaw({
                family: "vechain",
                estimatedFees: "210000000000000000",
                recipient: "0xcf130b42ae31c4931298b4b1c0f1d974b8732957",
                amount: "1000000000000000000",
                body: {
                  chainTag: MAINNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [
                    {
                      to: "0xcf130b42ae31c4931298b4b1c0f1d974b8732957",
                      value: "1000000000000000000",
                      data: "0x",
                    },
                  ],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                amount: new BigNumber("1000000000000000000"),
                estimatedFees: new BigNumber("210000000000000000"),
                totalSpent: new BigNumber("1000000000000000000"),
                errors: {},
                warnings: {},
              },
            },
            {
              name: "Send VTHO",
              transaction: fromTransactionRaw({
                family: "vechain",
                subAccountId:
                  "js:2:vechain:0x0fe6688548f0C303932bB197B0A96034f1d74dba:vechain+vechain%2Fvip180%2Fvtho",
                estimatedFees: "515180000000000000",
                recipient: "0xcf130b42ae31c4931298b4b1c0f1d974b8732957",
                amount: "1000000000000000000",
                body: {
                  chainTag: MAINNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [
                    {
                      to: "0x0000000000000000000000000000456e65726779",
                      value: 0,
                      data: VIP180.transfer.encode(
                        "0xcf130b42ae31c4931298b4b1c0f1d974b8732957",
                        "9000000000000000000",
                      ),
                    },
                  ],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                amount: new BigNumber("1000000000000000000"),
                estimatedFees: new BigNumber("515180000000000000"),
                totalSpent: new BigNumber("1515180000000000000"),
                errors: {},
                warnings: {},
              },
            },
            {
              name: "Amount required",
              transaction: fromTransactionRaw({
                family: "vechain",
                estimatedFees: "0",
                recipient: "0xcf130b42ae31c4931298b4b1c0f1d974b8732957",
                amount: "",
                body: {
                  chainTag: MAINNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [{ to: "", value: 0, data: "0x" }],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                errors: {
                  amount: new AmountRequired(),
                },
                warnings: {},
              },
            },
            {
              name: "VET balance not enough",
              transaction: fromTransactionRaw({
                family: "vechain",
                estimatedFees: "210000000000000000",
                recipient: "0xcf130b42ae31c4931298b4b1c0f1d974b8732957",
                amount: "20000000000000000000",
                body: {
                  chainTag: MAINNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [
                    {
                      to: "0xcf130b42ae31c4931298b4b1c0f1d974b8732957",
                      value: "20000000000000000000",
                      data: "0x",
                    },
                  ],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                amount: new BigNumber("20000000000000000000"),
                errors: {
                  amount: new NotEnoughBalance(),
                },
                warnings: {},
                totalSpent: new BigNumber("20000000000000000000"),
                estimatedFees: new BigNumber("210000000000000000"),
              },
            },
            {
              name: "VTHO balance not enough",
              transaction: fromTransactionRaw({
                family: "vechain",
                subAccountId:
                  "js:2:vechain:0x0fe6688548f0C303932bB197B0A96034f1d74dba:vechain+vechain%2Fvip180%2Fvtho",
                estimatedFees: "515820000000000000",
                recipient: "0xcf130b42ae31c4931298b4b1c0f1d974b8732957",
                amount: "20000000000000000000",
                body: {
                  chainTag: MAINNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [
                    {
                      to: "0xcf130b42ae31c4931298b4b1c0f1d974b8732957",
                      value: "20000000000000000000",
                      data: "0x",
                    },
                  ],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                amount: new BigNumber("20000000000000000000"),
                errors: {
                  amount: new NotEnoughBalance(),
                },
                warnings: {},
                totalSpent: new BigNumber("20515820000000000000"),
                estimatedFees: new BigNumber("515820000000000000"),
              },
            },
          ],
          FIXME_tests: [
            "balance is sum of ops", // the balance depends on VTHO and it's earned without operations
            "empty transaction is equals to itself", //nonce is not deterministic
            "ref stability on self transaction", //blockref is not deterministic
            "can be run in parallel and all yield same results", //blockref is not deterministic
          ],
        },
        {
          raw: vechain3 as AccountRaw,
          transactions: [
            {
              name: "Not enough VTHO to pay fees",
              transaction: fromTransactionRaw({
                family: "vechain",
                estimatedFees: "210000000000000000",
                recipient: "0xcf130b42ae31c4931298b4b1c0f1d974b8732957",
                amount: "1000000000000000000",
                body: {
                  chainTag: MAINNET_CHAIN_TAG,
                  blockRef: "0x00634a0c856ec1db",
                  expiration: 18,
                  clauses: [
                    {
                      to: "0xcf130b42ae31c4931298b4b1c0f1d974b8732957",
                      value: "1000000000000000000",
                      data: "0x",
                    },
                  ],
                  gasPriceCoef: DEFAULT_GAS_COEFFICIENT,
                  gas: "0",
                  dependsOn: null,
                  nonce: generateNonce(),
                },
              }),
              expectedStatus: {
                amount: new BigNumber("1000000000000000000"),
                estimatedFees: new BigNumber("210000000000000000"),
                totalSpent: new BigNumber("1000000000000000000"),
                errors: {
                  amount: new NotEnoughVTHO(),
                },
                warnings: {},
              },
            },
          ],
          FIXME_tests: [
            "balance is sum of ops", // the balance depends on VTHO and it's earned without operations
            "empty transaction is equals to itself", //nonce is not deterministic
            "ref stability on self transaction", //blockref is not deterministic
            "can be run in parallel and all yield same results", //blockref is not deterministic
          ],
        },
      ],
    },
  },
};

testBridge(dataset);
