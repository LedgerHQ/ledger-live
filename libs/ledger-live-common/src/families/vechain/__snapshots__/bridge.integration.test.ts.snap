// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`vechain currency bridge scanAccounts vechain seed 1 1`] = `
[
  {
    "currencyId": "vechain",
    "derivationMode": "vechain",
    "feesCurrencyId": "vechain/vip180/vtho",
    "freshAddress": "0x0fe6688548f0C303932bB197B0A96034f1d74dba",
    "freshAddressPath": "44'/818'/0'/0/0",
    "id": "js:2:vechain:0x0fe6688548f0C303932bB197B0A96034f1d74dba:vechain",
    "index": 0,
    "operationsCount": 1,
    "pendingOperations": [],
    "seedIdentifier": "04de5a8f11cc426ff5617aeffd397c5dd77b58424662f368f2cd539a8c42fd2e7d0c6245c9c3a4306f70d7f292e95642651a6832d66fd9823fa6a0af575113e94e",
    "subAccounts": [],
    "swapHistory": [],
    "syncHash": undefined,
    "used": true,
  },
  {
    "id": "js:2:vechain:0x0fe6688548f0C303932bB197B0A96034f1d74dba:vechain+vechain%2Fvip180%2Fvtho",
    "operationsCount": 1,
    "parentId": "js:2:vechain:0x0fe6688548f0C303932bB197B0A96034f1d74dba:vechain",
    "pendingOperations": [],
    "swapHistory": [],
    "tokenId": "vechain/vip180/vtho",
    "type": "TokenAccountRaw",
  },
  {
    "currencyId": "vechain",
    "derivationMode": "vechain",
    "feesCurrencyId": "vechain/vip180/vtho",
    "freshAddress": "0x02961B92B8D20A4ea12f1f1CeFA74Dd7B4355A86",
    "freshAddressPath": "44'/818'/0'/0/1",
    "id": "js:2:vechain:0x02961B92B8D20A4ea12f1f1CeFA74Dd7B4355A86:vechain",
    "index": 1,
    "operationsCount": 1,
    "pendingOperations": [],
    "seedIdentifier": "04de5a8f11cc426ff5617aeffd397c5dd77b58424662f368f2cd539a8c42fd2e7d0c6245c9c3a4306f70d7f292e95642651a6832d66fd9823fa6a0af575113e94e",
    "subAccounts": [],
    "swapHistory": [],
    "syncHash": undefined,
    "used": true,
  },
  {
    "id": "js:2:vechain:0x02961B92B8D20A4ea12f1f1CeFA74Dd7B4355A86:vechain+vechain%2Fvip180%2Fvtho",
    "operationsCount": 0,
    "parentId": "js:2:vechain:0x02961B92B8D20A4ea12f1f1CeFA74Dd7B4355A86:vechain",
    "pendingOperations": [],
    "swapHistory": [],
    "tokenId": "vechain/vip180/vtho",
    "type": "TokenAccountRaw",
  },
  {
    "currencyId": "vechain",
    "derivationMode": "vechain",
    "feesCurrencyId": "vechain/vip180/vtho",
    "freshAddress": "0xFbC8653C0b842dd33361CE8938B24e407dd4838C",
    "freshAddressPath": "44'/818'/0'/0/2",
    "id": "js:2:vechain:0xFbC8653C0b842dd33361CE8938B24e407dd4838C:vechain",
    "index": 2,
    "operationsCount": 0,
    "pendingOperations": [],
    "seedIdentifier": "04de5a8f11cc426ff5617aeffd397c5dd77b58424662f368f2cd539a8c42fd2e7d0c6245c9c3a4306f70d7f292e95642651a6832d66fd9823fa6a0af575113e94e",
    "subAccounts": [],
    "swapHistory": [],
    "syncHash": undefined,
    "used": false,
  },
  {
    "id": "js:2:vechain:0xFbC8653C0b842dd33361CE8938B24e407dd4838C:vechain+vechain%2Fvip180%2Fvtho",
    "operationsCount": 0,
    "parentId": "js:2:vechain:0xFbC8653C0b842dd33361CE8938B24e407dd4838C:vechain",
    "pendingOperations": [],
    "swapHistory": [],
    "tokenId": "vechain/vip180/vtho",
    "type": "TokenAccountRaw",
  },
]
`;

exports[`vechain currency bridge scanAccounts vechain seed 1 2`] = `
[
  [
    {
      "accountId": "js:2:vechain:0x0fe6688548f0C303932bB197B0A96034f1d74dba:vechain",
      "blockHash": "0x00fa5b35f158cee280b6be35b4ada7669951035a79806244aaa11e5d2764d5bd",
      "blockHeight": 16407349,
      "extra": {},
      "fee": "210000000000000000",
      "hash": "0xe113f82af98a24c6f2473802f8673f0d69eed1c0f51a5a87d3cc6b8cf96ccfd6",
      "id": "js:2:vechain:0x0fe6688548f0C303932bB197B0A96034f1d74dba:vechain-0xe113f82af98a24c6f2473802f8673f0d69eed1c0f51a5a87d3cc6b8cf96ccfd6-IN",
      "recipients": [
        "0x0fe6688548f0c303932bb197b0a96034f1d74dba",
      ],
      "senders": [
        "0xcf130b42ae31c4931298b4b1c0f1d974b8732957",
      ],
      "type": "IN",
      "value": "10000000000000000000",
    },
  ],
  [
    {
      "accountId": "js:2:vechain:0x0fe6688548f0C303932bB197B0A96034f1d74dba:vechain+vechain%2Fvip180%2Fvtho",
      "blockHash": "0x00fa5b4ee2eca5cf65907e3695bbea044cd1f78224534dc0ba02526890055ecc",
      "blockHeight": 16407374,
      "extra": {},
      "fee": "365180000000000000",
      "hash": "0x5fdd7191c4d476a8e86060d516366e87421f65667b9b3c14c33a740c04921b10",
      "id": "js:2:vechain:0x0fe6688548f0C303932bB197B0A96034f1d74dba:vechain+vechain%2Fvip180%2Fvtho-0x5fdd7191c4d476a8e86060d516366e87421f65667b9b3c14c33a740c04921b10-IN",
      "recipients": [
        "0x0fe6688548f0c303932bb197b0a96034f1d74dba",
      ],
      "senders": [
        "0xcf130b42ae31c4931298b4b1c0f1d974b8732957",
      ],
      "type": "IN",
      "value": "10000000000000000000",
    },
  ],
  [
    {
      "accountId": "js:2:vechain:0x02961B92B8D20A4ea12f1f1CeFA74Dd7B4355A86:vechain",
      "blockHash": "0x00fa5b547b0d542ba8e62a85aa99cf355f1b2a70bd34e3949f889b0856776e05",
      "blockHeight": 16407380,
      "extra": {},
      "fee": "210000000000000000",
      "hash": "0xa4096186f2d58ecc68e6ea11c9757ef5bd78b21b87d73b4618eb6fd3508431ec",
      "id": "js:2:vechain:0x02961B92B8D20A4ea12f1f1CeFA74Dd7B4355A86:vechain-0xa4096186f2d58ecc68e6ea11c9757ef5bd78b21b87d73b4618eb6fd3508431ec-IN",
      "recipients": [
        "0x02961b92b8d20a4ea12f1f1cefa74dd7b4355a86",
      ],
      "senders": [
        "0xcf130b42ae31c4931298b4b1c0f1d974b8732957",
      ],
      "type": "IN",
      "value": "10000000000000000000",
    },
  ],
  [],
  [],
  [],
]
`;
